subsys/llext/llext_handlers.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2024 Arduino SA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <zephyr/llext/llext.h>
+#include <zephyr/llext/loader.h>
+#include <zephyr/internal/syscall_handler.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(llext, CONFIG_LLEXT_LOG_LEVEL);
+
+#include "llext_priv.h"
+
+ssize_t z_impl_llext_get_init_array(struct llext *ext, void *buf, size_t buf_size)
+{
+	if (!ext) {
+		return -EINVAL;
+	}
+
+	size_t preinit_size = ext->mem_size[LLEXT_MEM_PREINIT];
+	size_t init_size = ext->mem_size[LLEXT_MEM_INIT];
+	size_t total_size = preinit_size + init_size;
+
+	if (buf) {
+		char *byte_ptr = buf;
+		const char **fn_ptrs = buf;
+
+		if (buf_size < total_size) {
+			return -ENOMEM;
+		}
+
+		memcpy(byte_ptr, ext->mem[LLEXT_MEM_PREINIT], preinit_size);
+		memcpy(byte_ptr + preinit_size, ext->mem[LLEXT_MEM_INIT], init_size);
+
+		/* Sanity check: init pointers must map inside the text region
+		 * of the extension. If this fails, something went wrong during
+		 * the relocation process.
+		 */
+		const char *text_start = ext->mem[LLEXT_MEM_TEXT];
+		const char *text_end = text_start + ext->mem_size[LLEXT_MEM_TEXT];
+
+		for (int i = 0; i < total_size / sizeof(void *); i++) {
+			if (fn_ptrs[i] < text_start || fn_ptrs[i] >= text_end) {
+				LOG_ERR("init function %i (%p) outside text region",
+					i, fn_ptrs[i]);
+				return -EFAULT;
+			}
+		}
+	}
+
+	return total_size;
+}
+
+#ifdef CONFIG_USERSPACE
+
+static int ext_is_valid(struct llext *ext, void *arg)
+{
+	return (ext == arg);
+}
+
+static inline ssize_t z_vrfy_llext_get_init_array(struct llext *ext, void *buf, size_t size)
+{
+	/* Test that ext matches a loaded extension */
+	K_OOPS(llext_iterate(ext_is_valid, ext) == 0);
+
+	if (buf) {
+		/* Test that buf is a valid user-accessible pointer */
+		K_OOPS(K_SYSCALL_MEMORY_WRITE(buf, size));
+	}
+
+	return z_impl_llext_get_init_array(ext, buf, size);
+}
+#include <zephyr/syscalls/llext_get_init_array_mrsh.c>
```

Comment:
Is this something automatically generated? I don't find this file - neither in the tree nor in this PR? But tests are passing, so I must be missing something

Reply:
Yes